lib/Slack_web_api.js

```diff
@@ -1,314 +1,180 @@
-                slack_api.callAPI('groups.rename', options, cb);
-            },
-            setPurpose: function(options, cb) {
-                slack_api.callAPI('groups.setPurpose', options, cb);
-            },
-            setTopic: function(options, cb) {
-                slack_api.callAPI('groups.setTopic', options, cb);
-            },
-            unarchive: function(options, cb) {
-                slack_api.callAPI('groups.unarchive', options, cb);
-            },
-        },
-        im: {
-            close: function(options, cb) {
-                slack_api.callAPI('im.close', options, cb);
-            },
-            history: function(options, cb) {
-                slack_api.callAPI('im.history', options, cb);
-            },
-            list: function(options, cb) {
-                slack_api.callAPI('im.list', options, cb);
-            },
-            mark: function(options, cb) {
-                slack_api.callAPI('im.mark', options, cb);
-            },
-            open: function(options, cb) {
-                slack_api.callAPI('im.open', options, cb);
-            }
-        },
-        mpim: {
-            close: function(options, cb) {
-                slack_api.callAPI('mpim.close', options, cb);
-            },
-            history: function(options, cb) {
-                slack_api.callAPI('mpim.history', options, cb);
-            },
-            list: function(options, cb) {
-                slack_api.callAPI('mpim.list', options, cb);
-            },
-            mark: function(options, cb) {
-                slack_api.callAPI('mpim.mark', options, cb);
-            },
-            open: function(options, cb) {
-                slack_api.callAPI('mpim.open', options, cb);
+
+                if (json.ok) {
+                    return cb(null, json);
+                }
+                return cb(json.error, json);
             }
-        },
-        pins: {
-            add: function(options, cb) {
-                slack_api.callAPI('pins.add', options, cb);
-            },
-            list: function(options, cb) {
-                slack_api.callAPI('pins.list', options, cb);
-            },
-            remove: function(options, cb) {
-                slack_api.callAPI('pins.remove', options, cb);
+            return cb(error || 'Invalid response');
+        }).form(options);
+    };
+
+
+    // generate all API methods
+    slackApiMethods.forEach(function(slackMethod) {
+        // most slack api methods are only two parts, of the form group.method, e.g. auth.test
+        // some have three parts: group.subgroup.method, e.g, users.profile.get
+        // this method loops through all groups in a method, ensures they exist,
+        // then adds the method to the terminal group
+
+        var groups = slackMethod.split('.');
+        var method = groups.pop();
+        var currentGroup = slack_api;
+        var nextGroupName;
+
+        for (var i = 0; i < groups.length; i++) {
```

Comment:
Using groups.length in a for loop will calculate the length of `groups` array in every iteration. (Because JavaScript sucks.) Please use lodash, `Array.forEach` or calculate the length of groups array before the loop.